### PR TITLE
response 404 for diff/patch of a commit that not exist

### LINKED ISSUE #15217: HTTP 500 when trying to download diff/patch of a commit that doesn't exist
- Gitea version (or commit ref): 1.13.6
- Git version: 2.20.1-2+deb10u3
- Operating system: Debian 10
  - I downloaded and installed Gitea using the instructions in [Installation from binary](https://docs.gitea.io/en-us/install-from-binary/).
  - Gitea is run by systemd
  
  
  
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
    - [Example URL](https://try.gitea.io/nguyena/code-server/commit/18ace7b9066c4492a61db4dbae3b49ad477891b4.diff) (I picked a random repo)
  - [ ] No
- Log gist: https://gist.github.com/BBaoVanC/f9c4c1275561a5dbd17f858468c2b19b

## Description

If you try to download the diff or patch of a commit that doesn't exist, then you get a 500. I would expect it to return a 404.

### PR DESCRIPTION
**Pull request by a1012112796**
fix #15217

### CODE REVIEW

**Review by lunny — APPROVED** [review]

File: routers/repo/commit.go [hunk 1]
```diff
@@ -388,6 +388,12 @@ func RawDiff(ctx *context.Context) {
 		git.RawDiffType(ctx.Params(":ext")),
 		ctx.Resp,
 	); err != nil {
+		if git.IsErrNotExist(err) {
+			ctx.JSON(404, map[string]interface{}{
+				"message": "commit " + ctx.Params(":sha") + " is not exist.",
```
Comment (silverwind) [inline routers/repo/commit.go]:
```suggestion
				"message": "commit " + ctx.Params(":sha") + " does not exist.",
```

File: routers/repo/commit.go [hunk 2]
```diff
@@ -388,6 +388,12 @@ func RawDiff(ctx *context.Context) {
 		git.RawDiffType(ctx.Params(":ext")),
 		ctx.Resp,
 	); err != nil {
+		if git.IsErrNotExist(err) {
+			ctx.JSON(404, map[string]interface{}{
+				"message": "commit " + ctx.Params(":sha") + " does not exist.",
+			})
+			return
```
Comment (zeripath) [inline routers/repo/commit.go]: GH just send a text reply  - what's the reasoning behind the json?

Comment (silverwind) [reply routers/repo/commit.go]: This is not API, right? Should probably send the 404 HTML page.

Comment (zeripath) [reply routers/repo/commit.go]: it's neither API nor html. it's a text endpoint for getting the diffs.

Comment (silverwind) [reply routers/repo/commit.go]: Hmm GH sends "Not Found" as `content-type: text/plain; charset=utf-8` in such a case, but if I go via browser, I still get a HTML page for some reason. Does not seem to be rely on user-agent sniffing thought.

```
$ curl -L https://github.com/go-gitea/gitea/pull/9999999.patch
Not Found%

Comment (silverwind) [reply routers/repo/commit.go]: Ah, they do it via the `accept` header. This gives HTML:

```
curl -L -H 'accept: text/html' https://github.com/go-gitea/gitea/pull/9999999.patch
```

Comment (a1012112796) [reply routers/repo/commit.go]: Hmm, so It's okay to response html page.  and we can change ``ctx.NotFound`` ro response different message type for different ``content-type`` request.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Please send backports

**Pull request merged**